internal/pagination/pagination.go

```diff
@@ -0,0 +1,261 @@
+
+	if len(items) > 0 {
+		resp.RefreshToken = refreshtoken.FromResource(items[len(items)-1], grantsHash)
+	}
+
+	return resp, nil
+}
+
+// ListRefresh returns a ListResponse. The response will contain at most a
+// number of items equal to the pageSize. Items are fetched using the
+// listRefreshItemsFn and then items are checked using the filterItemFn
+// to determine if they should be included in the response.
+// The response includes a new refresh token based on the grants and items.
+// The estimatedCountFn is used to provide an estimated total number of
+// items that can be returned by making additional requests using the provided
+// refresh token. The listDeletedIDsFn is used to list the IDs of any
+// resources that have been deleted since the refresh token was last used.
+func ListRefresh[T boundary.Resource](
+	ctx context.Context,
+	grantsHash []byte,
+	pageSize int,
+	filterItemFn ListFilterFunc[T],
+	listRefreshItemsFn ListRefreshItemsFunc[T],
+	estimatedCountFn EstimatedCountFunc,
+	listDeletedIDsFn ListDeletedIDsFunc,
+	tok *refreshtoken.Token,
+) (*ListResponse[T], error) {
+	const op = "pagination.ListRefresh"
+
+	if len(grantsHash) == 0 {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing grants hash")
+	}
+	if pageSize < 1 {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "page size must be at least 1")
+	}
+	if filterItemFn == nil {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing filter item callback")
+	}
+	if listRefreshItemsFn == nil {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing list refresh items callback")
+	}
+	if estimatedCountFn == nil {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing estimated count callback")
+	}
+	if listDeletedIDsFn == nil {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing list deleted IDs callback")
+	}
+	if tok == nil {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing refresh token")
+	}
+
+	deletedIds, transactionTimestamp, err := listDeletedIDsFn(ctx, tok.UpdatedTime)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	listItemsFn := func(ctx context.Context, prevPageLast T, limit int) ([]T, error) {
+		return listRefreshItemsFn(ctx, tok, prevPageLast, limit)
+	}
+
+	items, completeListing, err := list(ctx, grantsHash, pageSize, filterItemFn, listItemsFn)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	resp := &ListResponse[T]{
+		Items:           items,
+		CompleteListing: completeListing,
+		DeletedIds:      deletedIds,
+	}
+
+	resp.EstimatedItemCount, err = estimatedCountFn(ctx)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+
+	if len(items) > 0 {
+		resp.RefreshToken = tok.RefreshLastItem(items[len(items)-1], transactionTimestamp)
+	} else {
+		resp.RefreshToken = tok.Refresh(transactionTimestamp)
+	}
+
+	return resp, nil
+}
+
+func list[T boundary.Resource](
+	ctx context.Context,
+	grantsHash []byte,
```

Comment:
```suggestion
	_ []byte,
```
Since `grantsHash` is not used